### PR TITLE
Major ObjC Analysis Improvments

### PR DESCRIPTION
**Pull request by cxnder**
Official changelog:

- ObjC Processing now supports armv7+thumb binaries

- Type definition speedup

- Classes now have properly defined types

- ivars now render inline as struct members to 'self'

- off-image objc_msgSend calls now properly inline the selector string

---

devlog:

AbstractFile now has an `m_ptrSize` and `readPointer()` / `readPointer(uint64_t)` / `pointerSize() -> int`. These are used to make supporting across 32/64 bit trivial. Everything that reads a pointer (so basically every single `readLong()`) now uses this.

TypeParser now returns a `ParsedType` object, which holds the name + `std::opt<Ref<Type>>` + flavor (named, struct, etc) and an unused "dependency" field, which should(?) be used later in time to do dependency resolution when we add named types.

TypeParser now requires an `Ref<Architecture>` object as arg0 since it is generating pointer class Ref<Type>s and needs to know the proper pointer size.

CustomTypes no longer uses `ParseTypesFromSource()` for several reasons. All types in the project are now defined with the official and wonderfully documented C++ api for doing so. 

All types (except for specifically the `typedefs` detailed below) via the c++ api are now registered as system types, hiding them in the Types view by default. Perhaps even the typedefs should be hidden.

New `InfoHandler::CreateClassType()` which creates and registers a structure `class_ClassName` and a `typedef struct class_ClassName* ClassName`. It then registers all of the ivars into that struct using their hard coded offsets. This results in weird looking structs in the sidebar that have plenty of gaps, so we hide them, but renders perfectly in linearview/elsewhere.

Ivar list section now has its structs processed. 

All of the struct types are now defined based on the bv's address size and adjusted based on the official objc_runtime-old.h . 

In the initializing workflow pass, we now parse through all the msgSend symbols and define their type to `void *(* const _objc_msgSend)(id self, SEL sel)` which was not at all a headache to do with the C++ api. This causes all of the un-rewritten selectors to inline their sel strings, which is an insane improvement by itself.

---

Still a lot to do, but this is hopefully a good start. This PR resolves no open issues because I am too good for that.

### CODE REVIEW

**Comment by cxnder** [comment]
Hold off for a bit, obvious formatting issues + found a segfault.

**Comment by jonpalmisc** [comment]
Created #11 to track part of this. Will see how much this address other open issues during testing and link them accordingly.

File: Core/TypeParser.h
Comment (plafosse) [inline Core/TypeParser.h:38]: Architecture objects really don't need to be reference counted as they're singleton objects.

File: Workflow.cpp
Comment (plafosse) [inline Workflow.cpp:36]: whitespace

File: Workflow.cpp
Comment (plafosse) [inline Workflow.cpp:71]: whitespace

File: Workflow.cpp
Comment (plafosse) [inline Workflow.cpp:176]: be careful when assigning things to `void*` you should probably set the confidence to 0

**Review by plafosse — CHANGES_REQUESTED** [review]

File: Core/Analyzers/ClassAnalyzer.cpp
Comment (jonpalmisc) [inline Core/Analyzers/ClassAnalyzer.cpp:121]: These should follow the normal naming convention, e.g. `roPointerStart`, `roMethodOffset`, etc, etc.

File: Core/Analyzers/SelectorAnalyzer.cpp
Comment (jonpalmisc) [inline Core/Analyzers/SelectorAnalyzer.cpp:20]: Remove this blank line

File: Core/BinaryViewFile.cpp
Comment (jonpalmisc) [inline Core/BinaryViewFile.cpp:27]: This can be consolidated into `if (m_ptrSize == 4) readInt()` otherwise, `readLong()`

File: Core/BinaryViewFile.h
Comment (jonpalmisc) [inline Core/BinaryViewFile.h:25]: Please change to `m_pointerSize`

File: Core/AbstractFile.h
Comment (jonpalmisc) [inline Core/AbstractFile.h:64]: `s/arch/architecture/`, wrap comments at 80 columns

File: Core/AnalysisInfo.h
Comment (jonpalmisc) [inline Core/AnalysisInfo.h:133]: Remove added blank line

File: Core/TypeParser.cpp
Comment (jonpalmisc) [inline Core/TypeParser.cpp:47]: Remove blank line under `for` loop

File: Core/TypeParser.cpp
Comment (jonpalmisc) [inline Core/TypeParser.cpp:77]: Is this still a TODO?

File: InfoHandler.cpp
Comment (jonpalmisc) [inline InfoHandler.cpp:140]: Without looking too closely at what this is doing, my "need for a separate function" senses are going off. Seems to be a lot of indentation and complexity here which could benefit from either simplification or extraction into a helper function.

File: InfoHandler.cpp
Comment (jonpalmisc) [inline InfoHandler.cpp:234]: Remove braces since they are not used with single-line for loops in this project.

**Review by jonpalmisc — CHANGES_REQUESTED** [review]
# Formatting

There are plenty of formatting oddities and other issues; I believe I've left comments on all of the ones that _cannot_ be solved by `clang-format`.

The rest of the formatting inconsistencies can (and should) be solved by running `clang-format -i --style WebKit` on all sources and headers.

# Testing

The regression testing Objective Ninja formerly had was removed when it became an official plugin. I will look into adding it back in to ensure this PR doesn't introduce any unintended changes in analysis.

# Architecture

Objective Ninja was originally designed such that everything in the `Core` subdirectory had no dependency on Binary Ninja. The analysis used plain C++ and as long as you could provide your own valid implementation of `AbstractFile`, the analysis would work.

I haven't left comments on each individual violation of this, as there are numerous. One of two things would need to happen before I feel good about this being merged:

- Option A: `AbstractFile` is largely dissolved and Objective Ninja's usefulness as a backed-agnostic Objective-C analysis framework disappears.
- Option B: The changes presented here are restructured to allow the `Core` subdirectory to remain dependency-free and not be tied to Binary Ninja.

I would personally prefer Option B. In addition to retaining what I believe to be a nice quality of Objective Ninja, I also believe that having the core analysis and Binary Ninja-specific stuff separated make the code easier to reason about and helps maintain good separation of concerns.

**Comment by cxnder** [comment]
Closing this in favor of reworking the structure of the changes and tracking them as several smaller improvements vs one large mega-PR

**Pull request closed without merging**
